### PR TITLE
Refactoring utils to be more sane

### PR DESCRIPTION
**Pull request by artwr**
A refactor of the `airflow.utils` from a file into a folder with topics. As we scale out the Airflow code base, this will help us follow good practices of keeping small modules, and will make finding code easier. 

The goal of this PR is to establish foundations to that effect, following the examples set by projects like Django and IPython.

This is more of a starting point, and I welcome comments on how to make this better. I apologize for the sheer size fo the PR, the surgery was extensive. 

Note that this will include some backwards incompatible changes in `utils`. Since `utils` is mostly used by people extending Airflow, I believe this is still ok, and will help the community grow by making the code base better organized and easier to contribute to.

@mistercrunch

### CODE REVIEW

File: airflow/jobs.py [hunk 1]
```diff
@@ -308,12 +313,12 @@ def manage_slas(self, dag, session=None):
                 self.logger.info(' --------------> ABOUT TO CALL SLA MISS CALL BACK ')
                 dag.sla_miss_callback(dag, task_list, blocking_task_list, slas, blocking_tis)
                 notification_sent = True
-            from airflow import ascii
+            from airflow.utils import asciiart
```
Comment (aoen) [inline airflow/jobs.py]: Nit: Scope creep, but can you pull this to the top of the file?

File: airflow/models.py [hunk 2]
```diff
@@ -1238,7 +1242,7 @@ def email_alert(self, exception, is_retry=False):
             "Log file: {self.log_filepath}<br>"
             "Mark success: <a href='{self.mark_success_url}'>Link</a><br>"
         ).format(**locals())
-        utils.send_email(task.email, title, body)
+        utils.email.send_email(task.email, title, body)
```
Comment (aoen) [inline airflow/models.py]: Nit: from utils.email import send_email to be consistent with the other files

File: airflow/utils/classes.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+from __future__ import print_function
+from __future__ import unicode_literals
+
+from builtins import object
+
+import logging
+import signal
+
+
+class AirflowTaskTimeout(Exception):
```
Comment (aoen) [inline airflow/utils/classes.py]: I think you meant to remove this from here

File: airflow/utils/classes.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
+from __future__ import absolute_import
```
Comment (aoen) [inline airflow/utils/classes.py]: Nit: Do you think it would make sense to put the classes into their own files? better for modularization, and helps remove the vagueness of the name "classes.py"

File: airflow/utils/db.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+from alembic.config import Config
+from alembic import command
+from alembic.migration import MigrationContext
+
+from sqlalchemy import event, exc
+from sqlalchemy.pool import Pool
+
+from ..airflow import settings
+from ..airflow import configuration
```
Comment (aoen) [inline airflow/utils/db.py]: Could we make these absolute imports? http://stackoverflow.com/questions/4209641/absolute-vs-explicit-relative-import-of-python-module

**Comment by aoen** [comment]
Thanks for this change, this is great! Looks good, mostly nits.

File: airflow/utils/classes.py (hunk 3 above)
Comment (artwr) [inline airflow/utils/classes.py]: I was torn on this one since it is used right below in the timeout class. But I will move it to exceptions for consistency.

File: airflow/utils/classes.py (hunk 4 above)
Comment (artwr) [inline airflow/utils/classes.py]: I think it does. Actually, I would love to do the same to the `models.py`.

File: airflow/utils/classes.py (hunk 3 above)
Comment (aoen) [inline airflow/utils/classes.py]: I think it is in both places ATM.

**Comment by artwr** [comment]
That was epic. The build is now passing, let me know if people have more comments!

**Comment by aoen** [comment]
LGTM, thanks for addressing everything!

**Comment by mistercrunch** [comment]
It's just moving 1k lines around. What could go wrong?!

**Comment by aoen** [comment]
Here's hoping the test coverage is good ^^.

**Comment by artwr** [comment]
@mistercrunch The only potential issue that I have seen is the `apply_default` decorator. It now needs to be imported from `airflow.utils.decorators`, and so it depends how custom written operators have been written.

It should be a quick fix though.

**Comment by artwr** [comment]
@mistercrunch. Bump.

**Comment by mistercrunch** [comment]
Seems like `apply_default` should be also namespaced under `airflow.utils` or it will break all operators in the wild. We do not want to break all the operators that may exist, at least not until a big number version.

**Comment by artwr** [comment]
Superseded by https://github.com/airbnb/airflow/pull/1219
Closing this one.
